src/integer_mod_q/z_q/arithmetic/mul.rs

```diff
@@ -95,6 +99,49 @@ impl Zq {
 arithmetic_trait_borrowed_to_owned!(Mul, mul, Zq, Zq, Zq);
 arithmetic_trait_mixed_borrowed_owned!(Mul, mul, Zq, Zq, Zq);
 
+impl Mul<&Z> for &Zq {
+    type Output = Zq;
+
+    /// Implements the [`Mul`] trait for [`Zq`] and [`Z`] values.
+    /// [`Mul`] is implemented for any combination of owned and borrowed values.
+    ///
+    /// Parameters:
+    ///  - `other`: specifies the value to multiply with `self`
+    ///
+    /// Returns the product of both numbers as a [`Zq`].
+    ///
+    /// # Example
+    /// ```
+    /// use qfall_math::integer_mod_q::Zq;
+    /// use qfall_math::integer::Z;
+    /// use std::str::FromStr;
+    ///
+    /// let a: Zq = Zq::from_str("42 mod 19").unwrap();
+    /// let b: Z = Z::from_str("42").unwrap();
+    ///
+    /// let c: Zq = &a * &b;
+    /// let d: Zq = a * b;
+    /// let e: Zq = &c * Z::from(42);
+    /// let f: Zq = c * &Z::from(42);
+    /// ```
+    fn mul(self, other: &Z) -> Self::Output {
+        let mut out = fmpz(0);
+        unsafe {
+            fmpz_mod_mul_fmpz(
+                &mut out,
+                &self.value.value,
+                &other.value,
+                &*self.modulus.modulus,
+            );
+        }
+        Zq::from_z_modulus(&Z::from_fmpz(&out), &self.modulus)
```

Comment:
There two issues with this line:
1. `Zq::from_z_modulus` applies the modulus again (which is not necessary since `fmpz_mod_mul_fmpz` already ensures the reduced form.
2. `out` is never dropped. `Z::from_fmpz` does not take ownership, therefore this function would still be responsible for it.

```suggestion
        Zq {
            modulus: self.modulus.clone(),
            value: Z { value: out },
        }
```